coderd/database/queries/insights.sql

```diff
@@ -771,3 +771,124 @@ SELECT
 GROUP BY utp.num, utp.template_ids, utp.name, utp.type, utp.display_name, utp.description, utp.options, wbp.value;
+
+-- name: GetUserStatusCountsOverTime :many
+-- GetUserStatusCountsOverTime returns the count of users in each status over time.
+-- The time range is inclusively defined by the start_time and end_time parameters.
+--
+-- Bucketing:
+-- Between the start_time and end_time, we include each timestamp where a user's status changed or they were deleted.
+-- We do not bucket these results by day or some other time unit. This is because such bucketing would hide potentially
+-- important patterns. If a user was active for 23 hours and 59 minutes, and then suspended, a daily bucket would hide this.
+-- A daily bucket would also have required us to carefully manage the timezone of the bucket based on the timezone of the user.
+--
+-- Accumulation:
+-- We do not start counting from 0 at the start_time. We check the last status change before the start_time for each user. As such,
+-- the result shows the total number of users in each status on any particular day.
+WITH
+-- dates_of_interest defines all points in time that are relevant to the query.
+-- It includes the start_time, all status changes, all deletions, and the end_time.
+dates_of_interest AS (
+	SELECT @start_time::timestamptz AS date
+
+	UNION
+
+    SELECT DISTINCT changed_at AS date
+    FROM user_status_changes
+    WHERE changed_at > @start_time::timestamptz
+        AND changed_at < @end_time::timestamptz
+
+	UNION
+
+	SELECT DISTINCT deleted_at AS date
+	FROM user_deleted
+	WHERE deleted_at > @start_time::timestamptz
+		AND deleted_at < @end_time::timestamptz
+
+	UNION
+
+	SELECT @end_time::timestamptz AS date
+),
+-- latest_status_before_range defines the status of each user before the start_time.
+-- We do not include users who were deleted before the start_time. We use this to ensure that
+-- we correctly count users prior to the start_time for a complete graph.
+latest_status_before_range AS (
+    SELECT
+        DISTINCT usc.user_id,
+        usc.new_status,
+        usc.changed_at
+    FROM user_status_changes usc
+	LEFT JOIN user_deleted ud ON usc.user_id = ud.user_id
+    WHERE usc.changed_at < @start_time::timestamptz
+	AND (ud.user_id IS NULL OR ud.deleted_at > @start_time::timestamptz)
+    ORDER BY usc.user_id, usc.changed_at DESC
+),
+-- status_changes_during_range defines the status of each user during the start_time and end_time.
+-- If a user is deleted during the time range, we count status changes prior to the deletion.
+-- Theoretically, it should probably not be possible to update the status of a deleted user, but we
+-- need to ensure that this is enforced, so that a change in business logic later does not break this graph.
+status_changes_during_range AS (
+    SELECT
+        usc.user_id,
+        usc.new_status,
+        usc.changed_at
+    FROM user_status_changes usc
+	LEFT JOIN user_deleted ud ON usc.user_id = ud.user_id
```

Comment:
If a user is deleted twice in it's history (should not be possible, but would be nice to be durable against), then this join will duplicate the status changed rows for each deletion event.

We could make the table a subquery that returns 1 row, or you could drop a comment that if a user is deleted more than once, that breaks our logic.


Minimal example: https://www.db-fiddle.com/f/4jyoMCicNSZpjMt4jFYoz5/15847